lib/moonshot/command_line.rb

```diff
@@ -0,0 +1,108 @@
+require 'thor'
+
+module Moonshot
+  # This class implements the command-line `moonshot` tool.
+  class CommandLine
+    # TODO: No reason this is class-level, make this a for-real class that needs to be initialized.
+    def self.run!
+      # Find the Moonfile in this project.
+      orig_dir = Dir.pwd
+
+      loop do
+        break if File.exist?('Moonfile')
+
+        if Dir.pwd == '/'
+          warn 'Could not find Moonfile!'
+          exit 1
+        else
+          Dir.chdir('..')
+        end
+      end
+
+      moonfile_dir = Dir.pwd
+      Dir.chdir(orig_dir)
+
+      # Load any plugins and CLI extensions relative to the Moonfile
+      if File.directory?(File.join(moonfile_dir, 'moonshot'))
+        load_cli_extensions(moonfile_dir)
+        load_plugins(moonfile_dir)
+      end
+
+      Object.include(Moonshot::ArtifactRepository)
+      Object.include(Moonshot::BuildMechanism)
+      Object.include(Moonshot::DeploymentMechanism)
+      load(File.join(moonfile_dir, 'Moonfile'))
```

Comment:
We could support multiple names, like Rake does -- but I wonder if we should have a preference. I tend to agree that if this is ruby code it ends in .rb.